src/tree/binary_tree/mod.rs

```diff
@@ -0,0 +1,287 @@
+/// only enables the operations needed by MLS.
+#[derive(Clone, PartialEq, Debug, Serialize, Deserialize)]
+pub struct BinaryTree<T: PartialEq> {
+    nodes: Vec<T>,
+}
+
+impl<T: PartialEq> From<Vec<T>> for BinaryTree<T> {
+    fn from(nodes: Vec<T>) -> Self {
+        BinaryTree { nodes }
+    }
+}
+
+impl<T: PartialEq> BinaryTree<T> {
+    /// Get the size of the tree.
+    pub(crate) fn size(&self) -> NodeIndex {
+        NodeIndex::from(self.nodes.len())
+    }
+
+    /// Get the number of leaves in the tree.
+    pub(crate) fn leaf_count(&self) -> LeafIndex {
+        // We unwrap here, because we assume the tree to be full.
+        LeafIndex::try_from(self.size()).unwrap()
+    }
+
+    pub(crate) fn root(&self) -> NodeIndex {
+        let w = self.size();
+        NodeIndex::from((1usize << log2(w.as_usize())) - 1)
+    }
+
+    fn is_out_of_bounds(&self, index: NodeIndex) -> Result<(), BinaryTreeError> {
+        if index > self.size() {
+            return Err(BinaryTreeError::IndexOutOfBounds);
+        };
+        Ok(())
+    }
+
+    pub(crate) fn left(&self, index: NodeIndex) -> Result<NodeIndex, BinaryTreeError> {
+        self.is_out_of_bounds(index)?;
+        let x = index.as_usize();
+        let k = level(NodeIndex::from(x));
+        if k == 0 {
+            return Err(BinaryTreeError::LeafHasNoChildren);
+        }
+        Ok(NodeIndex::from(x ^ (0x01 << (k - 1))))
+    }
+
+    pub(crate) fn right(&self, index: NodeIndex) -> Result<NodeIndex, BinaryTreeError> {
+        self.is_out_of_bounds(index)?;
+        let size = self.leaf_count();
+        let x = index.as_usize();
+        let n = size.as_usize();
+        let k = level(NodeIndex::from(x));
+        if k == 0 {
+            return Err(BinaryTreeError::LeafHasNoChildren);
+        }
+        let mut r = x ^ (0x03 << (k - 1));
+        while r >= node_width(n) {
+            r = self.left(NodeIndex::from(r))?.as_usize();
+        }
+        Ok(NodeIndex::from(r))
+    }
+
+    pub(crate) fn parent(&self, index: NodeIndex) -> Result<NodeIndex, BinaryTreeError> {
+        self.is_out_of_bounds(index)?;
+        let size = self.leaf_count();
+        let x = index.as_usize();
+        let n = size.as_usize();
+        if index == self.root() {
+            return Err(BinaryTreeError::RootHasNoParent);
+        }
+        let mut p = parent_step(x);
+        while p >= node_width(n) {
+            p = parent_step(p)
+        }
+        Ok(NodeIndex::from(p))
+    }
+
+    pub(crate) fn sibling(&self, index: NodeIndex) -> Result<NodeIndex, BinaryTreeError> {
+        let p = self.parent(index)?;
+        match index.cmp(&p) {
+            Ordering::Less => self.right(p),
+            Ordering::Greater => self.left(p),
+            Ordering::Equal => self.left(p),
+        }
+    }
+
+    // Ordered from leaf to root
+    // Includes neither leaf nor root
+    pub(crate) fn dirpath(&self, index: NodeIndex) -> Result<Vec<NodeIndex>, BinaryTreeError> {
```

Comment:
These are still the old direct path functions I guess.

Reply:
Yes, I replaced them with the ones from your PR.